docs/throttle.md

```diff
@@ -0,0 +1,102 @@
+     valPower := entry.getValPower()
+     meter = meter - valPower
+     // Using the per-chain queue, handle the single slash packet using its queued data,
+     // then handle all trailing VSCMatured packets for this consumer
+     handleSlashPacketAndTrailingVSCMaturedPackets(entry)
+     // Delete entry in global queue, delete handled data
+     entry.Delete()
+     deletePendingSlashPacketData()
+     deleteTrailingVSCMaturedPacketData()
+}
+```
+
+## Endblocker handling - Slash Meter Replenishment
+
+Once the slash meter is not full, it'll be replenished after `SlashMeterReplenishPeriod (param)` by incrementing the meter by its allowance for that period, where `allowance` = `SlashMeterReplenishFraction (param)` * `currentTotalVotingPower`. The slash meter will never exceed its current allowance in value. Note a few things:
+
+1. The slash meter can go negative in value, and will do so when handling a single slash packet that jails a validator with significant voting power. In such a scenario, the slash meter may take multiple replenishment periods to once again reach a positive value, meaning no other slash packets may be handled for multiple replenishment periods.
+2. Total voting power of a chain changes over time, especially as validators are jailed/tombstoned. As validators are jailed, total voting power decreases, and so does the slashing allowance per replenishment period.
+3. The voting power allowance added to the slash meter per replenish period will always be strictly greater than 1. If the `SlashMeterReplenishFraction (param)` is set too low, integer rounding will put this minimum value into effect.
+
+The behavior described above is achieved by executing `CheckForSlashMeterReplenishment()` every endblock.
+
+## Throttling Invariant
+
+Using on-chain params and the sub protocol defined, slash packet throttling is implemented such that the following invariant is maintained (in addition to those already defined in the CCV spec).
+
+For the following invariant to hold, these points must be true:
+
+- We assume the total voting power of the chain (as a function of delegations) does not significantly increase over the course of the attack.
+- The final slashed validator does not have more than `SlashMeterReplenishFraction` of total voting power on the provider.
+- `SlashMeterReplenishFraction` is large enough to avoid rounding errors.
+- `SlashMeterReplenishPeriod` is sufficiently longer than the time it takes to produce a block.
+
+Invariant:
+
+> If we define a consumer initiated slash attack to start when the first slash packet from such an attack is received by the provider, and we define the initial validator set as the set that existed when the attack started, the time it takes to jail `X`% of the initial validator set will be greater than or equal to `(X * SlashMeterReplenishPeriod / SlashMeterReplenishFraction) - 2 * SlashMeterReplenishPeriod`
```

Comment:
> define a consumer initiated slash attack to start when the first slash packet from such an attack

For this to work as an invariant attack will also need to be defined.

Reply:
in spec, yes